### PR TITLE
#409 Aded UOM conversion check when adding a new price to a product.

### PR DESCRIPTION
**Pull request by dragospodariu96**
https://github.com/metasfresh/metasfresh/issues/6007

### CODE REVIEW

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

File: de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java [hunk 1]
```diff
@@ -124,6 +127,38 @@ public static void assertMainProductPriceIsNotDuplicate(final I_M_ProductPrice p
...
+		final IUOMConversionDAO uomConversionRepo = Services.get(IUOMConversionDAO.class);
+		final ProductId productId = ProductId.ofRepoId(productPrice.getM_Product_ID());
+
+		UOMConversionsMap conversionsMap = uomConversionRepo.getProductConversions(productId);
+
+		if (conversionsMap.getRateIfExists(UomId.ofRepoId(product.getC_UOM_ID()), UomId.ofRepoId(productPrice.getC_UOM_ID())).isPresent())
+		{
+			return;
+		}
+		throw new AdempiereException("UOM Conversion to the selected UOM doesn't exist").markAsUserValidationError();
```
Comment (metas-ts) [inline de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java]: Create an `AD_Message` record with translations (we'll need en_US, de_DE and de_CH) and use IMsg (`Services.get(IMsg.class)`) to get a translatable string instance.
Passing hardcoded strings  generally doesn't make sense for user-errors

Comment (teosarca) [reply de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java]: fully agree

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

File: de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java [hunk 2]
```diff
@@ -122,16 +132,6 @@ public static void assertMainProductPriceIsNotDuplicate(final I_M_ProductPrice p
 
 	public static void assertUomConversionExists(final I_M_ProductPrice productPrice)
 	{
-		if (productPrice == null || !productPrice.isActive())
```
Comment (metas-ts) [inline de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java]: i think it's not unnecessary. you never know who will call your `public static` method

File: de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java [hunk 3]
```diff
@@ -124,6 +130,29 @@ public static void assertMainProductPriceIsNotDuplicate(final I_M_ProductPrice p
 		getFirstOrThrowExceptionIfMoreThanOne(allMainPrices);
 	}
 
+	public static void assertUomConversionExists(final I_M_ProductPrice productPrice)
+	{
+		final IProductDAO productDAO = Services.get(IProductDAO.class);
```
Comment (metas-ts) [inline de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java]: why were the checks here unnecessary? remember, you never know who is going to call this `public static` method.
Maybe you can annotate the productPrice parameter with `@NonNull` (lombok!), but i thik you should check still for IsActive and IsValid

Comment (dragospodariu96) [reply de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java]: As it's only called in one place right now, it gets called at the same time as the assert before it (the one that checks that the product price is not duplicate), and the checks are removed are already gone through there.
But sure, for further use cases I could add them to be there.

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

**Comment by metas-ts** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- de.metas.business/src/main/java/de/metas/pricing/service/ProductPrices.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/metasfresh/metasfresh/pullRequest?prid=4783507&bid=15933219)

**Pull request merged**